typing/typecore.ml

```diff
@@ -1852,6 +1853,60 @@ struct
           true
     | _ -> false
 
+  let scrape env ty =
+    (Ctype.repr (Ctype.expand_head_opt env (Ctype.correct_levels ty))).desc
+
+let array_element_kind env ty =
+  match scrape env ty with
+  | Tvar _ | Tunivar _ ->
+      `Pgenarray
+  | Tconstr(p, _, _) ->
+      if Path.same p Predef.path_int || Path.same p Predef.path_char then
+        `Pintarray
+      else if Path.same p Predef.path_float then
+        `Pfloatarray
+      else if Path.same p Predef.path_string
+           || Path.same p Predef.path_array
+           || Path.same p Predef.path_nativeint
+           || Path.same p Predef.path_int32
+           || Path.same p Predef.path_int64 then
+        `Paddrarray
+      else begin
+        try
+          match Env'.find_type p env with
+            {type_kind = Type_abstract} ->
+              `Pgenarray
+          | {type_kind = Type_variant cstrs}
+            when List.for_all (fun c -> c.Types.cd_args = Types.Cstr_tuple [])
+                cstrs ->
+              `Pintarray
+          | {type_kind = _} ->
+              `Paddrarray
+        with Not_found ->
+          (* This can happen due to e.g. missing -I options,
+             causing some .cmi files to be unavailable.
+             Maybe we should emit a warning. *)
+          `Pgenarray
+      end
+  | _ ->
+      `Paddrarray
```

Comment:
Where does this code come from? Is it new? Can we trust that it corresponds to what the code generation is going to be doing?